### PR TITLE
#4434 Test Config Base Class

### PR DESCRIPTION
**Pull request by cj-dev**
Opening for review. If I can get confirmation this approach meets expectations outlined in the issue, I'll proceed with refactoring other tests that can take advantage of the `ConfigTestCase`.

### CODE REVIEW

**Comment by bmw** [comment]
This is largely on the right track with a couple of caveats:

1. Rather than using multiple inheritance in the test cases, you should just have the tests inherit from `ConfigTestCase` unless absolutely necessary. 
2. You probably want to use subdirectories of the temporary directory for the config, log, and work directories. I can see the directories interacting in ugly ways otherwise.
3. You may want to consider solving #4435 first. To build on my description in #4434, the config object is [built at runtime](https://github.com/certbot/certbot/blob/7d0e4d7bad7f602c63af01526d8b4d316e05e32e/certbot/main.py#L741) largely by just parsing the command line arguments. For testing, it'd be nice to have a config with all the default values (except a couple which may be modified to point to things like temporary directories). The reason this is nice is many tests duplicate the work of setting up these values so the tests run properly. If they don't, code may fail due to an attribute error or being unexpectedly passed a `mock` object somewhere down the line. Some tests may changes a value or two to not be the default, but the test changing it's copy of the config shouldn't be a problem and allowing the test to focus on what it's doing differently rather than defining boilerplate makes both reading and writing tests easier. #4435 which I just added more information to describes a way to create this default config. While I'd take a PR that duplicates the keys and values in the config here, it'd be cleaner and more maintainable in the long run to use the approach specified in #4435 to help solve this PR.

Happy to answer any questions if you have them.

**Comment by cj-dev** [comment]
1. Oversight on my part. Will do.
2. Absolutely, makes sense.
3. I'll take a jab at that too and keep you posted on progress.

File: certbot/tests/util.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 from certbot import util
 
 from certbot.display import util as display_util
+from certbot.configuration import NamespaceConfig
```
Comment (bmw) [inline certbot/tests/util.py]: nit: Our [style guide](https://certbot.eff.org/docs/contributing.html#coding-style) recommends [only importing packages/modules](https://google.github.io/styleguide/pyguide.html#Imports) so it'd be better to just import `configuration` and use ` configuration.NamespaceConfig`.

File: certbot/tests/account_test.py [hunk 2]
```diff
@@ -98,14 +93,13 @@ def test_it(self):
...
-class AccountFileStorageTest(TempDirTestCase):
+class AccountFileStorageTest(test_util.ConfigTestCase):
     """Tests for certbot.account.AccountFileStorage."""
 
     def setUp(self):
         super(AccountFileStorageTest, self).setUp()
 
-        self.config = mock.MagicMock(
-            accounts_dir=os.path.join(self.tempdir, "accounts"))
+        self.config.accounts_dir = os.path.join(self.tempdir, "accounts")
```
Comment (bmw) [inline certbot/tests/account_test.py]: We actually don't need this line as `NamespaceConfig` defines `accounts_dir` relative to `config_dir` ([source](https://github.com/certbot/certbot/blob/edcfc49303d4ddd2e69ba6ac6933af01d00e332c/certbot/configuration.py#L67)).

File: certbot/tests/cert_manager_test.py [hunk 3]
```diff
@@ -18,58 +18,52 @@
...
-        os.makedirs(os.path.join(self.tempdir, "renewal"))
-
-        self.cli_config = configuration.NamespaceConfig(mock.MagicMock(
-            config_dir=self.tempdir,
-            work_dir=self.tempdir,
-            logs_dir=self.tempdir,
-            quiet=False,
-        ))
+        self.config.quiet = False
+        # TODO: This and others should use the NamespaceConfig properties, no?
```
Comment (bmw) [inline certbot/tests/cert_manager_test.py]: Yes that's a good idea. For example, it'd be `self.config.renewal_configs_dir` below.

File: certbot/tests/cert_manager_test.py [hunk 4]
```diff
@@ -354,89 +340,92 @@ def _call(self, *args, **kwargs):
...
-            self._call, mock_config)
+            self._call, self.config)
 
     @test_util.patch_get_utility()
     @mock.patch("certbot.storage.RenewableCert._check_symlinks")
     def test_rename_cert(self, mock_check, unused_get_utility):
         mock_check.return_value = True
-        mock_config = self.mock_config
-        self._call(mock_config)
+        # mock_config = self.mock_config
```
Comment (bmw) [inline certbot/tests/cert_manager_test.py]: Can we delete this line?

File: certbot/tests/cert_manager_test.py [hunk 5]
```diff
@@ -354,89 +340,92 @@ def _call(self, *args, **kwargs):
...
-        self.assertEqual(updated_lineage.lineagename, mock_config.new_certname)
+        self.assertEqual(updated_lineage.lineagename, self.config.new_certname)
 
     @test_util.patch_get_utility()
     @mock.patch("certbot.storage.RenewableCert._check_symlinks")
     def test_rename_cert_interactive_certname(self, mock_check, mock_get_utility):
         mock_check.return_value = True
-        mock_config = self.mock_config
-        mock_config.certname = None
+        # mock_config = self.mock_config
```
Comment (bmw) [inline certbot/tests/cert_manager_test.py]: Can we delete this line?

File: certbot/tests/cert_manager_test.py [hunk 6]
```diff
@@ -354,89 +340,92 @@ def _call(self, *args, **kwargs):
...
         self.assertTrue(updated_lineage is not None)
-        self.assertEqual(updated_lineage.lineagename, mock_config.new_certname)
+        self.assertEqual(updated_lineage.lineagename, self.config.new_certname)
 
     @test_util.patch_get_utility()
     @mock.patch("certbot.storage.RenewableCert._check_symlinks")
     def test_rename_cert_bad_new_certname(self, mock_check, unused_get_utility):
         mock_check.return_value = True
-        mock_config = self.mock_config
+        # mock_config = self.mock_config
```
Comment (bmw) [inline certbot/tests/cert_manager_test.py]: Can we delete this line?

File: certbot/tests/main_test.py [hunk 7]
```diff
@@ -811,36 +804,37 @@ def test_quiet_renew(self):
...
-        args = mock.MagicMock(account=None, config_dir=self.config_dir,
-                              logs_dir=self.logs_dir, work_dir=self.work_dir,
-                              email=None, webroot_path=None)
-        config = configuration.NamespaceConfig(args)
-        lineage = storage.RenewableCert(rc_path, config)
-        renewalparams = lineage.configuration["renewalparams"]
-        # pylint: disable=protected-access
-        renewal._restore_webroot_config(config, renewalparams)
-        self.assertEqual(config.webroot_path, ["/var/www/"])
+    # Duplicated in renewal_test.py
```
Comment (bmw) [inline certbot/tests/main_test.py]: Nice catch! Instead of commenting these lines out, can we delete them?

**Comment by cj-dev** [comment]
Thanks for the feedback! I still do intend to come back around to 4435, but I need to put a ribbon on that other issue we've been working on.

**Pull request merged**